cohortextractor/main.py

```diff
@@ -32,18 +37,102 @@ def generate_cohort(
         definition_path=definition_path,
         output_file=output_file,
         backend=backend_id,
+        index_date_range=index_date_range,
         dummy_data_file=dummy_data_file,
     )
 
-    cohort = load_cohort(definition_path)
     output_file.parent.mkdir(parents=True, exist_ok=True)
-    if dummy_data_file and not db_url:
-        validate_dummy_data(cohort, dummy_data_file, output_file)
-        shutil.copyfile(dummy_data_file, output_file)
+
+    for index_date in _generate_date_range(index_date_range):
+        if index_date is not None:
+            log.info(f"Setting index_date to {index_date}")
+            date_suffix = f"_{index_date}"
+        else:
+            date_suffix = ""
+        cohort = load_cohort(definition_path, index_date)
+        output_file_with_date = Path(str(output_file).replace("*", date_suffix))
+        if dummy_data_file and not db_url:
+            dummy_data_file_with_date = Path(
+                str(dummy_data_file).replace("*", date_suffix)
+            )
+            validate_dummy_data(
+                cohort, dummy_data_file_with_date, output_file_with_date
+            )
+            shutil.copyfile(dummy_data_file_with_date, output_file_with_date)
+        else:
+            backend = BACKENDS[backend_id](
+                db_url, temporary_database=temporary_database
+            )
+            results = extract(cohort, backend)
+            write_output(results, output_file_with_date)
+
+
+def _generate_date_range(date_range_str):
+    # Bail out with an "empty" range: this means we don't need separate
+    # codepaths to handle the range, single date, and no date supplied cases
+    if not date_range_str:
+        return [None]
+    start, end, period = _parse_date_range(date_range_str)
+    if end < start:
+        raise ValueError(
+            f"Invalid date range '{date_range_str}': end cannot be earlier than start"
+        )
+    dates = []
+    while start <= end:
+        dates.append(start.isoformat())
+        start = _increment_date(start, period)
+    # The latest data is generally more interesting/useful so we may as well
+    # extract that first
+    dates.reverse()
```

Comment:
This seems _so_ unlikely to be genuinely useful. I bet it originates from some earlier, unreliable version of the system where just getting _something_ out was considered a wild succcess. @evansd?

Reply:
I think that, as with the `--keep-existing-files` option, this originates from pre-job-runner days where output was going to a researcher's working directory on TPP. Under those circumstances you could start inspecting output before the extraction had finished running and so this made more sense. As Ben says, it's pointless now.

Reply:
OK, will ditch it